Source/ASNetworkImageNode.mm

```diff
@@ -617,43 +655,49 @@ - (void)_lazilyLoadImageIfNecessary
         }
 
         if (imageContainer != nil) {
-          strongSelf->_imageLoaded = YES;
+          [strongSelf _locked_setCurrentImageQuality:1.0];
           if ([imageContainer asdk_animatedImageData] && _downloaderFlags.downloaderImplementsAnimatedImage) {
-            strongSelf.animatedImage = [_downloader animatedImageWithData:[imageContainer asdk_animatedImageData]];
+            id animatedImage = [_downloader animatedImageWithData:[imageContainer asdk_animatedImageData]];
+            [strongSelf _locked_setAnimatedImage:animatedImage];
           } else {
-            [strongSelf _setImage:[imageContainer asdk_image]];
+            [strongSelf _locked__setImage:[imageContainer asdk_image]];
           }
-          dispatch_async(dispatch_get_main_queue(), ^{
-            strongSelf->_currentImageQuality = 1.0;
-          });
+          strongSelf->_imageLoaded = YES;
         }
 
         strongSelf->_downloadIdentifier = nil;
-
         strongSelf->_cacheUUID = nil;
 
         if (imageContainer != nil) {
-          if (strongSelf->_delegateFlags.delegateDidLoadImage) {
-            [strongSelf->_delegate imageNode:strongSelf didLoadImage:strongSelf.image];
+          if (_delegateFlags.delegateDidLoadImage) {
+            ASDN::MutexUnlocker u(strongSelf->__instanceLock__);
+            [delegate imageNode:strongSelf didLoadImage:strongSelf.image];
           }
         }
-        else if (error && strongSelf->_delegateFlags.delegateDidFailWithError) {
-          [strongSelf->_delegate imageNode:strongSelf didFailWithError:error];
+        else if (error && _delegateFlags.delegateDidFailWithError) {
+          ASDN::MutexUnlocker u(strongSelf->__instanceLock__);
+          [delegate imageNode:strongSelf didFailWithError:error];
         }
       };
 
       if (_cache != nil) {
```

Comment:
Nit: Might worth a comment that explains why we don't need the lock to access `_cache`, i.e The ivar is set once in the initializer.